# Estimating Labor Market Returns to Education

In this exercise, we're going to use data from the [American Communities Survey (ACS)](https://usa.ipums.org/usa/acs.shtml) to study the relationship betwen educational attainment and wages. The ACS is a survey conducted by the United States Census Bureau (though it is not "The Census," which is a counting of every person in the United States that takes place every 10 years) to measure numerous features of the US population. The data we will be working with includes about 100 variables from the 2017 ACS survey, and is a 10% sample of the ACS (which itself is a 1% sample of the US population, so we're working with about a 0.1% sample of the United States). 

This data comes from [IPUMS](https://usa.ipums.org/usa/), which provides a very useful tool for getting subsets of major survey datasets, not just from the US, but [from government statistical agencies the world over](https://international.ipums.org/international-action/sample_details).

This is *real* data, meaning that you are being provided the data as it is provided by IPUMS. Documentation for all variables used in this data can be found [here](https://usa.ipums.org/usa-action/variables/group) (you can either search by variable name to figure out the meaning of a variable in this data, or search for something you want to see if a variable with the right name is in this data). 

Within this data is information on both the educational background and current earnings of a representative sample of Americans. We will now use this data to estimate the labor-market returns to graduating high school and college, and to learn something about the meaning of an educational degree. 

## Gradescope Autograding

Please follow [all standard guidance](https://www.practicaldatascience.org/html/autograder_guidelines.html) for submitting this assignment to the Gradescope autograder, including storing your solutions in a dictionary called `results` and ensuring your notebook runs from the start to completion without any errors.

For this assignment, please name your file `exercise_dataframes.ipynb` before uploading.

You can check that you have answers for all questions in your `results` dictionary with this code:

```python
assert set(results.keys()) == {
    "ex2_num_obs",
    "ex3_num_vars",
    "ex8_updated_num_obs",
    "ex9_updated_num_obs",
    "ex11_grade12_income",
    "ex12_college_income",
    "ex12_college_income_pct",
    "ex14_high_school_dropout",
    "ex15_grade_9",
    "ex15_grade_10",
    "ex15_grade_11",
    "ex15_grade_12",
    "ex15_4_years_of_college",
    "ex15_graduate",
}
```

### Submission Limits

Please remember that you are **only allowed three submissions to the autograder.** Your last submission (if you submit 3 or fewer times), or your third submission (if you submit more than 3 times) will determine your grade Submissions that error out will **not** count against this total.


In [25]:
import pandas as pd
import numpy as np

pd.set_option("mode.copy_on_write", True)

# Create a results dictionary
results = {}

## Exercises

### Exercise 1

Data for these [exercises can be found here](https://github.com/nickeubank/MIDS_Data/tree/master/US_AmericanCommunitySurvey). 

Import `US_ACS_2017_10pct_sample.dta` into a pandas DataFrame (read it directly from a URL to help the autograder, please). 

This can be done with the command `pd.read_stata`, which will read in files created in the program Stata (and which uses the file suffix `.dta`). This is a format commonly used by social scientists.

In [26]:
import pandas as pd

# URL of the dataset US_ACS_2017_10pct_sample.dta
url = "https://github.com/nickeubank/MIDS_Data/raw/master/US_AmericanCommunitySurvey/US_ACS_2017_10pct_sample.dta"

# Reading the data from the URL into a pandas DataFrame
acs = pd.read_stata(url)

## Getting to Know Your Data

When you get a new dataset like this, it's good to start by trying to get a feel for its contents and organization. Toy datasets you sometimes get in classes are often very small, and easy to look at, but this is a pretty large dataset, so you can't just open it up and get a good sense of it. Here are some ways to get to know your data. 

### Exercise 2

How many observations are in your data? Store the answer in your `results` dictionary with the key `"ex2_num_obs"`.

In [27]:
# df.shape[0] returns the number of rows in df
ex2_num_obs = acs.shape[0]
results["ex2_num_obs"] = ex2_num_obs
print(f"The number of observations in my data is {ex2_num_obs}.")

The number of observations in my data is 319004.


### Exercise 3

How many variables are in your data? Store the answer in your `results` dictionary with the key `"ex3_num_vars"`.

In [28]:
# df.shape[1] returns the number of columns in df
ex3_num_vars = acs.shape[1]
results["ex3_num_vars"] = ex3_num_vars

print(f"The number of variables in my data is {ex3_num_vars}.")

The number of variables in my data is 104.


### Exercise 4

 Let's see what variables are in this dataset. First, try to see them all using the command:


```python
acs.columns
```

In [29]:
acs.columns

Index(['year', 'datanum', 'serial', 'cbserial', 'numprec', 'subsamp', 'hhwt',
       'hhtype', 'cluster', 'adjust',
       ...
       'migcounty1', 'migmet131', 'vetdisab', 'diffrem', 'diffphys', 'diffmob',
       'diffcare', 'diffsens', 'diffeye', 'diffhear'],
      dtype='object', length=104)

As you will see, `python` doesn't like to print out all the different variables when there are this many in a dataset. 

To get everything printed out, we can loop over all the columns and print them one at a time with the command:

```
for c in acs.columns: print(c)
```

It's definitely a bit of a hack, but honestly a pretty useful one!

In [30]:
for c in acs.columns:
    print(c)

year
datanum
serial
cbserial
numprec
subsamp
hhwt
hhtype
cluster
adjust
cpi99
region
stateicp
statefip
countyicp
countyfip
metro
city
citypop
strata
gq
farm
ownershp
ownershpd
mortgage
mortgag2
mortamt1
mortamt2
respmode
pernum
cbpernum
perwt
slwt
famunit
sex
age
marst
birthyr
race
raced
hispan
hispand
bpl
bpld
citizen
yrnatur
yrimmig
language
languaged
speakeng
hcovany
hcovpriv
hinsemp
hinspur
hinstri
hcovpub
hinscaid
hinscare
hinsva
hinsihs
school
educ
educd
gradeatt
gradeattd
schltype
degfield
degfieldd
degfield2
degfield2d
empstat
empstatd
labforce
occ
ind
classwkr
classwkrd
looking
availble
inctot
ftotinc
incwage
incbus00
incss
incwelfr
incinvst
incretir
incsupp
incother
incearn
poverty
migrate1
migrate1d
migplac1
migcounty1
migmet131
vetdisab
diffrem
diffphys
diffmob
diffcare
diffsens
diffeye
diffhear


### Exercise 5

That's a *lot* of variables, and definitely more than we need. In general, life is easier when working with these kinds of huge datasets if you can narrow down the number of variables a little. In this exercise, we will be looking at the relationship between education and wages, we need variables for: 

- Age
- Income
- Education
- Employment status (is the person actually working)

These quantities of interest correspond to the following variables in our data: `age`, `inctot`, `educ`, and `empstat`. 

Subset your data to just those variables. 

In [31]:
# Create a subset dataframe of acs including Age, Income, Education, Employment status
selected_columns = ["age", "inctot", "educ", "empstat"]
subset_acs = acs[selected_columns].copy()

### Exercise 6 

Now that we have a more manageable number of variables, it's often very useful to look at a handful of rows of your data. The easiest way to do this is probably the `.head()` method (which will show you the first five rows), or the `tail()` method, which will show you the last five rows. 

But to get a good sense of your data, it's often better to use the `sample()` command, which returns a random set of rows. As the first and last rows are sometimes not representative, a random set of rows can be very helpful. Try looking at a random sample of 20 rows (note: you don't have to run `.sample()` ten times to get ten rows. Look at the `.sample` help file if you're stuck. 

In [32]:
# Display the first few rows of the subset dataframe
subset_acs.head()

age   inctot                       educ             empstat
0   4  9999999  nursery school to grade 4                 n/a
1  17     6000                   grade 11            employed
2  63     6150         4 years of college            employed
3  66    14000                   grade 12  not in labor force
4   1  9999999        n/a or no schooling                 n/a

In [33]:
# Display the last few rows of the subset dataframe
subset_acs.tail()

age   inctot                       educ   empstat
318999  33    22130         4 years of college  employed
319000   4  9999999  nursery school to grade 4       n/a
319001  20     5000                   grade 12  employed
319002  47   240000        5+ years of college  employed
319003  33    48000        5+ years of college  employed

In [34]:
# Returns a random set of rows from the subset dataframe
subset_acs.sample()

age  inctot      educ             empstat
274989  24    3000  grade 12  not in labor force

In [35]:
# Returns a random set of 20 rows from the subset dataframe
subset_acs.sample(n=20, random_state=1)

age   inctot                       educ             empstat
290939  20     1400          1 year of college  not in labor force
124459  83     6800        grade 5, 6, 7, or 8  not in labor force
190165  54   265000        5+ years of college            employed
178595  27    25000          1 year of college            employed
166299  39    32600                   grade 12            employed
303755  54     1300                    grade 9  not in labor force
251810  16        0                    grade 9  not in labor force
82814   74    13100                   grade 12  not in labor force
214301  59    60000                   grade 12            employed
249702  10  9999999  nursery school to grade 4                 n/a
241455  60    22100          1 year of college            employed
134451  45      130                   grade 12  not in labor force
41924   12  9999999        grade 5, 6, 7, or 8                 n/a
157130  94    14400                   grade 12  not in labor force
76589   44    80000          1 year of college            employed
87015   55     7400                   grade 12            employed
278613  68     8900          1 year of college  not in labor force
180674  68    22500          1 year of college  not in labor force
87924   11  9999999        grade 5, 6, 7, or 8                 n/a
90323   56        0         4 years of college  not in labor force

### Exercise 7

Do you see any immediate problems? What issues do you see? (Please do answer in markdown)

> I observed that individuals approximately around the age of 10, presumably young children not participating in the workforce, have been recorded with unusually high incomes, approximately around $9,999,999. This seems abnormal and may represent erroneous data entries.

### Exercise 8 

One problem is that many people seem to have incomes of $9,999,999. Moreover, people with those incomes seem to be very young children. 

What you are seeing is one method (a relatively old one) for representing missing data. In this case, the value 9999999 is being used as a **sentinel value** — a way to denote missing data that was used back in the day when there was no way to add a special data type for mossing data. In this case, it identifies observations where the person is too young to work, so their income value is missing. 

So let's begin by dropping anyone who has `inctot` equal to 9999999.

After dropping, how many observations do you have? Save your answer in your `results` dictionary under the key `"ex8_updated_num_obs"`

In [36]:
# Getting the indices of the rows where 'inctot' is 9999999
drop_index = subset_acs[subset_acs["inctot"] == 9999999].index

# Dropping those rows from the DataFrame
filtered_subset_acs = subset_acs.drop(drop_index)

# Counting the number of observations after filtering
ex8_updated_num_obs = len(filtered_subset_acs)

results["ex8_updated_num_obs"] = len(filtered_subset_acs)

# Printing the results
print(
    f"The number of the obsrvations after dropping income equal to $9,999,999 is {ex8_updated_num_obs}."
)

print(results)

The number of the obsrvations after dropping income equal to $9,999,999 is 265103.
{'ex2_num_obs': 319004, 'ex3_num_vars': 104, 'ex8_updated_num_obs': 265103}


### Exercise 9

OK, the other potential problem is that our data includes lots of people who are unemployed and people who are not in the labor force (this means they not only don't have a job, but also aren't looking for a job). For this analysis, we want to focus on the wages of people who are currently employed. So subset the dataset for the people for whom `empstat` is equal to "employed". 

Note that our decision to only look at people who are employed impacts how we should interpret the relationship we estimate between education and income. Because we are only looking at employed people, we will be estimating the relationship between education and income *for people who are employed*. That means that if education affects the *likelihood* someone is employed, we won't capture that in this analysis.

(You might also want to run `.sample()` after this just to make sure you were successful in your subsetting).

After this subsetting, how many observations do you have? Save your answer in your `results` dictionary under the key `"ex9_updated_num_obs"`

In [37]:
# Subsetting filtered_subset_acs to include only rows where 'empstat' is "employed"
employed_subset_acs = filtered_subset_acs[filtered_subset_acs["empstat"] == "employed"]

# Checking a sample of the subsetted DataFrame to verify the subsetting
employed_subset_acs.sample(n=20, random_state=1)

age  inctot                 educ   empstat
34744   57   15200    1 year of college  employed
231975  23   30000   4 years of college  employed
185443  63   85000   4 years of college  employed
285814  40   28600             grade 12  employed
45685   44   38000             grade 12  employed
81932   51   70000             grade 12  employed
169563  28   46800             grade 12  employed
91185   35   31600             grade 12  employed
205801  37   20000   2 years of college  employed
308950  29   17500             grade 12  employed
133207  49   26000             grade 12  employed
215442  36   68000  5+ years of college  employed
193624  60   45000   2 years of college  employed
95110   69   29200  5+ years of college  employed
198687  25   21000    1 year of college  employed
69448   64   43000             grade 12  employed
308015  54   34900   2 years of college  employed
82716   48   56500   2 years of college  employed
22534   59    3000  5+ years of college  employed
29821   68   46200    1 year of college  employed

In [38]:
# Counting the number of observations after subsetting
ex9_updated_num_obs = len(employed_subset_acs)

results["ex9_updated_num_obs"] = ex9_updated_num_obs

# Printing the number of observations left after subsetting for employed individuals
print(
    f"The number of observations left after subsetting for employed individuals is {ex9_updated_num_obs}."
)

print(results)

The number of observations left after subsetting for employed individuals is 148758.
{'ex2_num_obs': 319004, 'ex3_num_vars': 104, 'ex8_updated_num_obs': 265103, 'ex9_updated_num_obs': 148758}


### Exercise 10

Now let's turn to education. The `educ` variable seems to have a lot of discrete values. Let's see what values exist, and their distribution, using the `value_counts()` method. This is an *extremely* useful tool you'll use a lot! Try the following code (modified for the name of your dataset, of course):

```python
acs["educ"].value_counts()
```

In [39]:
# Checking the unique values and their counts in the 'educ' column of employed_subset_acs DataFrame
educ_value_counts = employed_subset_acs["educ"].value_counts()

# Printing the value counts for 'educ' column
print(educ_value_counts)

educ
grade 12                     47815
4 years of college           33174
1 year of college            22899
5+ years of college          20995
2 years of college           14077
grade 11                      2747
grade 5, 6, 7, or 8           2092
grade 10                      1910
n/a or no schooling           1291
grade 9                       1290
nursery school to grade 4      468
Name: count, dtype: int64


### Exercise 11

There are a lot of values in here, so let's just check a couple. What is the average value of `inctot` for people whose highest grade level is "grade 12" (in the US, that is someone who has graduated high school)?

Save your answer in your `results` dictionary under the key `"ex11_grade12_income"`.

In [40]:
# Calculate the average income for people who has only graduated high school
ex11_grade12_income = employed_subset_acs[employed_subset_acs["educ"] == "grade 12"][
    "inctot"
].mean()

results["ex11_grade12_income"] = ex11_grade12_income

print(
    f"The average income for employed high school gradautes is about ${round(ex11_grade12_income,2)}."
)

print(results)

The average income for employed high school gradautes is about $38957.76.
{'ex2_num_obs': 319004, 'ex3_num_vars': 104, 'ex8_updated_num_obs': 265103, 'ex9_updated_num_obs': 148758, 'ex11_grade12_income': 38957.76068179442}


### Exercise 12

What is the average income of someone who has completed an undergraduate degree but not done any postgraduate education ("4 years of college")? 

Save your answer in your `results` dictionary under the key `"ex12_college_income"`.

In percentage terms, how much does an employed college graduate earn as compared to someone who is only a high school graduate? Use the reference category that gives an answer above 100.

Store your answer in `"ex12_college_income_pct"`. Put your answer in percentage terms (so 100 implies they earn the same amount).

*Make sure to interpret your result in words when you print it out!*

In [41]:
# Calculate the average income for people who has only completed an undergraduate degree
ex12_college_income = employed_subset_acs[
    employed_subset_acs["educ"] == "4 years of college"
]["inctot"].mean()

results["ex12_college_income"] = ex12_college_income

# the amount an employed college graduate earn as compared to someone who is only a high school graduate
ex12_college_income_pct = (
    (1 + (ex12_college_income - ex11_grade12_income) / ex11_grade12_income)
) * 100

results["ex12_college_income_pct"] = ex12_college_income_pct

# Printing the results
print(
    f"The average income for employed individuals who have completed an undergraduate degree is about ${round(ex12_college_income, 2)}."
)
print(
    f"The amount an employed college graduate earn as compared to someone who is only a high school graduate is about {round(ex12_college_income_pct, 2)} % in percentage terms."
)
print(
    f"This means that the employed college graduate earns about {round(ex12_college_income_pct-100, 2)} % more compared to someone who is a employed high school graduate on average."
)
print(
    "This suggests that obtaining an undergraduate degree substantially increases one’s earning potential compared to holding only a high school diploma."
)

print(results)

The average income for employed individuals who have completed an undergraduate degree is about $75485.05.
The amount an employed college graduate earn as compared to someone who is only a high school graduate is about 193.76 % in percentage terms.
This means that the employed college graduate earns about 93.76 % more compared to someone who is a employed high school graduate on average.
This suggests that obtaining an undergraduate degree substantially increases one’s earning potential compared to holding only a high school diploma.
{'ex2_num_obs': 319004, 'ex3_num_vars': 104, 'ex8_updated_num_obs': 265103, 'ex9_updated_num_obs': 148758, 'ex11_grade12_income': 38957.76068179442, 'ex12_college_income': 75485.05293301983, 'ex12_college_income_pct': 193.7612727527617}



### Exercise 13
What does that suggest is the value of getting a college degree after graduating high school?

> + A 93.76% increase in income for employed college graduates, compared to employed high school graduates on average, suggests that obtaining an undergraduate degree substantially increases one’s earning potential compared to holding only a high school diploma.
> + This 93.76% increase is significant, showing the substantial economic value and benefits of higher education. 
> + This data demonstrates a strong correlation between higher education and increased income.

### Exercise 14

What is the average income for someone who has not finished high school? What does that suggest is the value of a high school diploma? (Treat `n/a or no schooling` as having no formal schooling, not as missing).

**Hint:** You may find the [.isin()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.isin.html) method to be really helpful here.

Save your answer in your `results` dictionary under the key `"ex14_high_school_dropout"`.

In [42]:
# Selecting rows where educ indicates not having finished high school
high_school_not_finished = [
    "n/a or no schooling",
    "nursery school to grade 4",
    "grade 5, 6, 7, or 8",
    "grade 9",
    "grade 10",
    "grade 11",
]

# Filter the employed_subset_acs for high school drop out
dropout_subset = employed_subset_acs[
    employed_subset_acs["educ"].isin(high_school_not_finished)
]

# Calculate the average income for high school drop out
ex14_high_school_dropout = dropout_subset["inctot"].mean()

results["ex14_high_school_dropout"] = ex14_high_school_dropout

print(
    f"The average income for someone who has not finished high school is about ${round(ex14_high_school_dropout, 2)}"
)

print(results)

The average income for someone who has not finished high school is about $26226.46
{'ex2_num_obs': 319004, 'ex3_num_vars': 104, 'ex8_updated_num_obs': 265103, 'ex9_updated_num_obs': 148758, 'ex11_grade12_income': 38957.76068179442, 'ex12_college_income': 75485.05293301983, 'ex12_college_income_pct': 193.7612727527617, 'ex14_high_school_dropout': 26226.45692998571}


### Exercise 15 

Complete the following table (storing values under the provided keys where listed):

- Average income for someone who only completed 9th grade (`ex15_grade_9`): _________
- Average income for someone who only completed 10th grade (`ex15_grade_10`): _________
- Average income for someone who only completed 11th grade (`ex15_grade_11`): _________
- Average income for someone who finished high school (12th grade) but never started college (`ex15_grade_12`): _________
- Average income for someone who completed 4 year of college (in the US, this corresponds to getting an undergraduate degree), but has no post-graduate education (no more than 4 years, `ex15_4_years_of_college`): _________
- Average income for someone who has some graduate education (more than 4 years, `ex15_graduate`): _________

In [43]:
# Define a calculate_average_income function to calculate and store average income for a given education level
def calculate_average_income(educ_level, key):
    avg_income = employed_subset_acs[employed_subset_acs["educ"] == educ_level][
        "inctot"
    ].mean()
    results[key] = avg_income
    print(
        f"Average income for someone who has an education level of {educ_level} is about ${round(avg_income, 2)}"
    )


# Average income for someone who only completed 9th grade
calculate_average_income("grade 9", "ex15_grade_9")

# Average income for someone who only completed 10th grade
calculate_average_income("grade 10", "ex15_grade_10")

# Average income for someone who only completed 11th grade
calculate_average_income("grade 11", "ex15_grade_11")

# Average income for someone who finished high school but never started college
calculate_average_income("grade 12", "ex15_grade_12")

# Average income for someone who completed 4 years of college but has no postgraduate education
calculate_average_income("4 years of college", "ex15_4_years_of_college")

# Average income for someone who has some graduate education
calculate_average_income("5+ years of college", "ex15_graduate")

print(results)

Average income for someone who has an education level of grade 9 is about $27171.91
Average income for someone who has an education level of grade 10 is about $23018.8
Average income for someone who has an education level of grade 11 is about $21541.69
Average income for someone who has an education level of grade 12 is about $38957.76
Average income for someone who has an education level of 4 years of college is about $75485.05
Average income for someone who has an education level of 5+ years of college is about $110013.22
{'ex2_num_obs': 319004, 'ex3_num_vars': 104, 'ex8_updated_num_obs': 265103, 'ex9_updated_num_obs': 148758, 'ex11_grade12_income': 38957.76068179442, 'ex12_college_income': 75485.05293301983, 'ex12_college_income_pct': 193.7612727527617, 'ex14_high_school_dropout': 26226.45692998571, 'ex15_grade_9': 27171.907751937986, 'ex15_grade_10': 23018.795811518325, 'ex15_grade_11': 21541.68693119767, 'ex15_grade_12': 38957.76068179442, 'ex15_4_years_of_college': 75485.05293301

### Exercise 16 

Why do you think there is no benefit from moving from grade 9 to grade 10, or grade 10 to grade 11, but there is a huge benefit to moving from grade 11 to graduating high school (grade 12)?

(Think carefully before reading ahead!)

> + I believe this is because many positions list a high school diploma as a baseline qualification requirement. It is often considered the minimum requirement for numerous jobs. 
> + Those positions without high school diploma as a minimum requirement are less paid. 
> + A high school diploma is an excellent demonstration of a candidate's learning ability and the necessary skills to succeed in a position.
> + Thus, there might not be substantial income differences between those who completed up to grade 9, 10, or 11, but a significant difference does exist between individuals who have obtained a high school degree and those who have not.

In [44]:
assert set(results.keys()) == {
    "ex2_num_obs",
    "ex3_num_vars",
    "ex8_updated_num_obs",
    "ex9_updated_num_obs",
    "ex11_grade12_income",
    "ex12_college_income",
    "ex12_college_income_pct",
    "ex14_high_school_dropout",
    "ex15_grade_9",
    "ex15_grade_10",
    "ex15_grade_11",
    "ex15_grade_12",
    "ex15_4_years_of_college",
    "ex15_graduate",
}

## Take-aways

Congratulations! You just discovered "the sheepskin effect!": people with degrees tend to earn substantially more than people who have *almost* as much education, but don't have an actual degree. 

In economics, this is viewed as evidence that the reason employers pay people with high school degrees more than those without degree is *not* that they think those who graduated high school have learned specific, useful skills. If that were the case, we would expect employee earnings to rise with every year of high school, since in each year of high school we learn more. 

Instead, this suggests employees pay high school graduates more because they think *the kind of people* who can finish high school are the *kind of people* who are likely to succeed at their jobs. Finishing high school, in other words, isn't about accumulating specific knowledge; it's about showing that you *are the kind of person* who can rise to the challenge of finishing high school, also suggesting you are also the kind of person who can succeed as an employee. 

(Obviously, this does not tell us whether that is an *accurate* inference, just that that seems to be how employeers think.) 

In other words, in the eyes of employers, a high school degree is a *signal* about the kind of person you are, not certification that you've learned a specific set of skills (an idea that earned [Michael Spence](https://en.wikipedia.org/wiki/Michael_Spence) a Nobel Prize in Economics). 